server/client.go

```diff
@@ -535,24 +728,99 @@ func (c *client) maxPayloadViolation(sz int, max int64) {
 	c.closeConnection()
 }
 
-// Assume the lock is held upon entry.
-func (c *client) sendProto(info []byte, doFlush bool) error {
-	var err error
-	if c.bw != nil && c.nc != nil {
-		deadlineSet := false
-		if doFlush || c.bw.Available() < len(info) {
-			c.nc.SetWriteDeadline(time.Now().Add(c.srv.getOpts().WriteDeadline))
-			deadlineSet = true
+// queueOutbound queues data for client/route connections.
+// Return pending length.
+// Lock should be held.
+func (c *client) queueOutbound(data []byte) {
+	// Add to pending bytes total.
+	c.out.pb += int64(len(data))
+
+	// Check for slow consumer via pending bytes limit.
+	// ok to return here, client is going away.
+	if c.out.pb > c.out.mp {
+		c.clearConnection()
+		atomic.AddInt64(&c.srv.slowConsumers, 1)
+		c.Noticef("Slow Consumer Detected: MaxPending of %d Exceeded", c.out.mp)
+		return
+	}
+
+	if c.out.p == nil && len(data) < maxBufSize {
+		if c.out.sz == 0 {
+			c.out.sz = startBufSize
+		}
+		if c.out.s != nil && cap(c.out.s) >= c.out.sz {
+			c.out.p = c.out.s
+			c.out.s = nil
+		} else {
+			// FIXME(dlc) - make power of 2 if less than maxBufSize?
+			c.out.p = make([]byte, 0, c.out.sz)
 		}
-		_, err = c.bw.Write(info)
-		if err == nil && doFlush {
-			err = c.bw.Flush()
+	}
+	// Determine if we copy or reference
+	available := cap(c.out.p) - len(c.out.p)
+	if len(data) > available {
+		// We can fit into existing primary, but message will fit in next one
+		// we allocate or utilize from the secondary. So copy what we can.
+		if available > 0 && len(data) < c.out.sz {
+			c.out.p = append(c.out.p, data[:available]...)
+			data = data[available:]
 		}
-		if deadlineSet {
-			c.nc.SetWriteDeadline(time.Time{})
+		// Put the primary on the nb if it has a payload
+		if len(c.out.p) > 0 {
+			c.out.nb = append(c.out.nb, c.out.p)
+			c.out.p = nil
+		}
+		// Check for a big message, and if found place directly on nb
+		// FIXME(dlc) - do we need signaling of ownership here if we want len(data) <
+		if len(data) > maxBufSize {
+			c.out.nb = append(c.out.nb, data)
```

Comment:
What happens if the data is referenced here? What I mean is for instance when delivering a message, we queue the message header and then the payload. But the header is already referencing a client's buffer that is reused for each deliverMsg call. Is that safe?

Reply:
Good observation. The code above is safe since anything over maxBufSize is a heap allocation for the message, hence the conditional. I want to make it more explicit such that I could scale out fan out cpu wise. Ran out of time but still on my list. I could add some more comments about how it is safe now.